### PR TITLE
AA8K-9, AA8K-8, AA8K-3, AA8K-2 - Rewrite routes and Healthchecks to Typescript

### PR DESCRIPTION
**Pull request by KacperMalachowski**
Rewrited routes, app initialization and healthchecks to typescript. 

Added changes from master and connection to MongoDB.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,22 +1,52 @@
...
+  },
+  "lint-staged": {
+    "*.{js,ts}": [
+      "npm run lint",
+      "npm run test"
+    ]
+  },
   "dependencies": {
-    "body-parser": "^1.19.0",
+    "@types/express": "^4.17.7",
```
Comment (hayuna) [inline package.json]: Types should be moved to devDependencies

File: package.json [hunk 2]
```diff
@@ -1,22 +1,52 @@
...
+    "cors": "^2.8.5",
     "dotenv": "^8.2.0",
     "express": "^4.17.1",
-    "mongoose": "^5.9.28"
+    "mongoose": "^5.9.28",
+    "passport": "^0.4.1",
+    "passport-github": "^1.1.0"
   },
   "devDependencies": {
+    "@types/passport": "^1.0.4",
```
Comment (hayuna) [inline package.json:44]: To better organisation, you should use some reordering eslint rules, it should be alphabetic order

File: package.json [hunk 3]
```diff
@@ -1,22 +1,52 @@
...
+    "@typescript-eslint/eslint-plugin": "^3.10.1",
+    "@typescript-eslint/parser": "^3.10.1",
+    "eslint": "^7.7.0",
+    "eslint-config-prettier": "^6.11.0",
+    "eslint-plugin-prettier": "^3.1.4",
+    "lint-staged": "^10.2.13",
+    "prettier": "^2.1.1",
+    "body-parser": "^1.19.0",
```
Comment (hayuna) [inline package.json]: Body-parser is not longer need as it was moved to express

File: package.json [hunk 4]
```diff
@@ -1,22 +1,52 @@
...
+    "dotenv": "^8.2.0",
+    "express": "^4.17.1",
```
Comment (hayuna) [inline package.json]: express should be moved to dependencies

File: package.json [hunk 5]
```diff
@@ -1,22 +1,52 @@
...
+    "mongoose": "^5.9.28",
```
Comment (hayuna) [inline package.json]: express should be moved to dependencies

File: src/app.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { Express, default as express } from 'express';
+import cors from 'cors';
+import routes from './routes';
+
+export default function initialize(app: Express): Express {
```
Comment (hayuna) [inline src/app.ts]: Can we replace this function into arrow function?

File: src/config/passport.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  (accessToken, refreshToken, profile, cb) => {
+    User
+      .findOneOrCreate(accessToken, profile)
+      .then((user) => {
+        if(user) 
+          cb(null, user);
+        else
+          cb();
+      })
+}))
```
Comment (hayuna) [inline src/config/passport.js]: I think we should have any exception block here `catch`

File: src/config/passport.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+passport.serializeUser((user, done) => done(null, user.id));
+
+passport.deserializeUser((userID, done) => {
```
Comment (hayuna) [inline src/config/passport.js]: Check if this is possible to use async await instead of callback in whole file

File: src/routes/health.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { Router, Request, Response } from 'express';
+import mongoose from 'mongoose';
+import HealthCheckController from '../controllers/HealthCheckController';
```
Comment (hayuna) [inline src/routes/health.ts:3]: We can use absolute paths in TS

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+import initialize from './app';
+import { connect } from './connectToMongo';
+
+const { MONGO_CONNECTION_STRING } = process.env as {
+  [key: string]: string;
+};
+
+connect(MONGO_CONNECTION_STRING);
+
+const app = express();
```
Comment (hayuna) [inline src/index.ts]: I'm curious we need this variable here. Could we move it to method initialize? Then we can remove additional parameter in this method

File: src/controllers/HealthCheckController.ts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        isReady: this.isConnected(),
+        status: this.isConnected() ? 1 : 0,
+        message: this.serviceStatusMessage(),
+      },
+    };
+  }
+
+  private serviceStatusMessage() {
+    switch (this.connection.readyState) {
+      case 0 || 2 || 3:
```
Comment (hayuna) [inline src/controllers/HealthCheckController.ts]: :o Never seen this signature, is it working?

File: src/controllers/HealthCheckController.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+      case 3:
+        return ' Disconecting from database...';
+
+      default:
+        return 'Unknown database connection state!';
+    }
+  }
+
+  private isConnected() {
+    return this.connection.readyState == 1;
```
Comment (hayuna) [inline src/controllers/HealthCheckController.ts]: Use triple equals

File: src/controllers/HealthCheckController.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+  public getServiceStatus(): IResult {
+    return {
+      database: {
+        isConnected: this.isConnected(),
+        status: this.connection.readyState,
+        message: this.databaseStatusMessage(),
+      },
+      service: {
+        isReady: this.isConnected(),
+        status: this.isConnected() ? 1 : 0,
```
Comment (hayuna) [inline src/controllers/HealthCheckController.ts:28]: Would be better if we return true/false in this case? Maybe you can use BANG operator to do this

File: src/__tests__/HealthCheckController.spec.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
+import HealthCheckController from '../controllers/HealthCheckController';
+import mongoose from 'mongoose';
+
+jest.mock('mongoose');
+let connection: mongoose.Connection;
```
Comment (hayuna) [inline src/__tests__/HealthCheckController.spec.ts]: Definitely should not be `let` because this variable never changed

**Review by hayuna — CHANGES_REQUESTED** [review]
Comments above

File: src/controllers/HealthCheckController.ts (hunk 13 above)
Comment (KacperMalachowski) [reply src/controllers/HealthCheckController.ts:28]: isReady is boolean. status is numeric version of isReady for feature better description current service status,

File: src/config/passport.js (hunk 8 above)
Comment (KacperMalachowski) [reply src/config/passport.js]: No, passport allow only callback method.

File: src/__tests__/HealthCheckController.spec.ts (hunk 14 above)
Comment (KacperMalachowski) [reply src/__tests__/HealthCheckController.spec.ts]: Rewrited: Now connection is set once on test run

File: src/controllers/HealthCheckController.ts (hunk 12 above)
Comment (KacperMalachowski) [reply src/controllers/HealthCheckController.ts]: Changed to triple equals

File: src/controllers/HealthCheckController.ts (hunk 11 above)
Comment (KacperMalachowski) [reply src/controllers/HealthCheckController.ts]: Changed to something better to case stack.
case 0:
case 2:
case 3:

File: src/index.ts (hunk 10 above)
Comment (KacperMalachowski) [reply src/index.ts]: Moved to initialize

File: src/routes/health.ts (hunk 9 above)
Comment (KacperMalachowski) [reply src/routes/health.ts]: Add tsconfig configuration to use it, and changed paths to absolute where was necessary.

Comment (KacperMalachowski) [reply src/routes/health.ts:3]: It's not working for me correctly.

File: package.json [hunk 15]
```diff
@@ -1,35 +1,63 @@
...
+    "lint": "eslint src/**/*.[jt]s",
+    "lint:fix": "eslint src/**/*.[jt]s --fix",
+    "start": "node dist/index.js",
     "test": "jest",
     "test:watch": "jest --watch"
   },
-  "keywords": [],
-  "author": "",
-  "license": "ISC",
+  "husky": {
```
Comment (hayuna) [inline package.json:17]: Now we have this configuration twice implemented (line 17 and line 58) move it to one `husky` property

Comment (KacperMalachowski) [reply package.json:17]: I check for that situations, but I miss it. Moved

**Review by hayuna — DISMISSED** [review]

File: src/config/index.js [hunk 16]
```diff
@@ -1,5 +1,6 @@
+require('dotenv').config();
```
Comment (Przemocny) [inline src/config/index.js]: why it in js, not in ts

File: src/config/passport.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
+const passport = require('passport');
+const GithubStrategy = require('passport-github').Strategy;
+const { User } = require('../models/User');
+const config = require('./index');
```
Comment (Przemocny) [inline src/config/passport.js]: import > require and convert to typescript

File: src/controllers/HealthCheckController.ts [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        message: this.serviceStatusMessage(),
+      },
+    };
+  }
+
+  private serviceStatusMessage() {
```
Comment (Przemocny) [inline src/controllers/HealthCheckController.ts]: missing returned type private serviceStatusMessage() : string

File: src/index.ts [hunk 19]
```diff
@@ -0,0 +1,17 @@
+import dotenv from 'dotenv';
+import initialize from '@/app';
+import { connect } from '@/connectToMongo';
+
+dotenv.config();
+
+const { MONGO_CONNECTION_STRING } = process.env as {
+  [key: string]: string;
+};
```
Comment (Przemocny) [inline src/index.ts:9]: there should be global process type declaration or either 

interface IProcessEnv { ... }

File: src/routes/health.ts [hunk 20]
```diff
@@ -0,0 +1,14 @@
+import { Router, Request, Response } from 'express';
+import mongoose from 'mongoose';
+import HealthCheckController from '@/controllers/HealthCheckController';
```
Comment (Przemocny) [inline src/routes/health.ts]: imo should be @controllers not @/controllers

**Review by Przemocny — CHANGES_REQUESTED** [review]
little changes, all should be in TS

File: src/config/index.js (hunk 16 above)
Comment (KacperMalachowski) [reply src/config/index.js]: It's old configuration. Task AA8K-14 was created to rewrite it to process.env

File: src/config/passport.js (hunk 17 above)
Comment (KacperMalachowski) [reply src/config/passport.js]: It's converted on branch AA8K-13. I created PR to this branch for this change.

File: src/routes/health.ts (hunk 20 above)
Comment (KacperMalachowski) [reply src/routes/health.ts]: Both ways doesn't work.

File: src/app.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+  app.use(express.urlencoded({ extended: true }));
+  app.use(express.json());
+  app.use(cors());
+
+  app.use(passport.initialize());
+  app.use(passport.session());
+
+  app.use(routes());
+
+  app.use(passport.initialize());
```
Comment (hayuna) [inline src/app.ts:20]: We can see duplicates

**Pull request closed without merging**